x-pack/auditbeat/processors/add_session_metadata/processdb/db.go

```diff
@@ -0,0 +1,742 @@
+			db.createEntryLeader(pid, EntryConsole)
+			db.logger.Debugf("entry_eval %d: entry type is console", p.Pids.Tgid)
+			return &pid
+		case p.Pids.Ppid == 1:
+			db.createEntryLeader(pid, Init)
+			db.logger.Debugf("entry_eval %d: entry type is init", p.Pids.Tgid)
+			return &pid
+		case !isFilteredExecutable(procBasename):
+			if parent, ok := db.processes[p.Pids.Ppid]; ok {
+				parentBasename := basename(parent.Filename)
+				if ttyType == Pts && parentBasename == "ssm-session-worker" {
+					db.createEntryLeader(pid, Ssm)
+					db.logger.Debugf("entry_eval %d: entry type is ssm", p.Pids.Tgid)
+					return &pid
+				} else if parentBasename == "sshd" && procBasename != "sshd" {
+					// TODO: get ip from env vars
+					db.createEntryLeader(pid, Sshd)
+					db.logger.Debugf("entry_eval %d: entry type is sshd", p.Pids.Tgid)
+					return &pid
+				} else if isContainerRuntime(parentBasename) {
+					db.createEntryLeader(pid, Container)
+					db.logger.Debugf("entry_eval %d: entry type is container", p.Pids.Tgid)
+					return &pid
+				}
+			}
+		default:
+			db.logger.Debugf("entry_eval %d: is a filtered executable: %s", p.Pids.Tgid, procBasename)
+		}
+	}
+
+	// if not a session leader or was not determined to be an entry leader, get
+	// it via parent, session leader, group leader (in that order)
+	relations := []struct {
+		pid  uint32
+		name string
+	}{
+		{
+			pid:  p.Pids.Ppid,
+			name: "parent",
+		},
+		{
+			pid:  p.Pids.Sid,
+			name: "session_leader",
+		},
+		{
+			pid:  p.Pids.Pgid,
+			name: "group_leader",
+		},
+	}
+
+	for _, relation := range relations {
+		if entry, ok := db.entryLeaderRelationships[relation.pid]; ok {
+			entryType := db.entryLeaders[entry]
+			db.logger.Debugf("entry_eval %d: got entry_leader: %d (%s), from relative: %d (%s)", p.Pids.Tgid, entry, string(entryType), relation.pid, relation.name)
+			return &entry
+		} else {
+			db.logger.Debugf("entry_eval %d: failed to find relative: %d (%s)", p.Pids.Tgid, relation.pid, relation.name)
+		}
+	}
+
+	// if it's a session leader, then make it its own entry leader with unknown
+	// entry type
+	if p.Pids.Tgid == p.Pids.Sid {
+		db.createEntryLeader(pid, EntryUnknown)
+		db.logger.Debugf("entry_eval %d: this is a session leader and no relative has an entry leader. entry type is unknown", p.Pids.Tgid)
+		return &pid
+	}
+
+	db.logger.Debugf("entry_eval %d: this is not a session leader and no relative has an entry leader, entry_leader will be unset", p.Pids.Tgid)
+	return nil
+}
+
+func (db *DB) InsertSetsid(setsid types.ProcessSetsidEvent) error {
+	db.Lock()
+	defer db.Unlock()
+
+	if entry, ok := db.processes[setsid.Pids.Tgid]; ok {
+		entry.Pids = pidInfoFromProto(setsid.Pids)
+		db.processes[setsid.Pids.Tgid] = entry
+	} else {
+		db.processes[setsid.Pids.Tgid] = Process{
+			Pids: pidInfoFromProto(setsid.Pids),
+		}
+	}
+
+	return nil
+}
+
+func (db *DB) InsertExit(exit types.ProcessExitEvent) error {
```

Comment:
Why does this have an error return in the signature? It always returns nil.

Reply:
yep, the return isn't needed. I've removed it from all the similar public functions